stko/optimizers/collapser.py

```diff
@@ -228,60 +231,571 @@ def optimize(self, mol):
+                f"====================\n"
+            )
+        return mol
+
+
+class CollapserMC(Collapser):
+    """
+    Collapse molecule to decrease enlarged bonds using MC algorithm.
+
+    Smarter optimisation than Collapser using simple Monte Carlo
+    algorithm to perform rigid translations of building blocks.
+
+    """
+
+    def __init__(
+        self,
+        output_dir,
+        step_size,
+        target_bond_length,
+        num_steps,
+        bond_epsilon=50,
+        nonbond_epsilon=20,
+        nonbond_sigma=1.2,
+        nonbond_mu=3,
+        beta=2,
+    ):
+        """
+        Initialize a :class:`Collapser` instance.
+
+        Parameters
+        ----------
+        output_dir : :class:`str`
+            The name of the directory into which files generated during
+            the calculation are written, if ``None`` then
+            :func:`uuid.uuid4` is used.
+
+        step_size : :class:`float`
+            The relative size of the step to take during step.
+
+        target_bond_length : :class:`float`
+            Target equilibrium bond length for long bonds to minimize
+            to.
+
+        num_steps : :class:`int`
+            Number of MC moves to perform.
+
+        bond_epsilon : :class:`float`, optional
+            Value of epsilon used in the bond potential in MC moves.
+            Determines strength of the bond potential.
+            Defaults to 50.
+
+        nonbond_epsilon : :class:`float`, optional
+            Value of epsilon used in the nonbond potential in MC moves.
+            Determines strength of the nonbond potential.
+            Defaults to 20.
+
+        nonbond_sigma : :class:`float`, optional
+            Value of sigma used in the nonbond potential in MC moves.
+            Defaults to 1.2.
+
+        nonbond_mu : :class:`float`, optional
+            Value of mu used in the nonbond potential in MC moves.
+            Determines the steepness of the nonbond potential.
+            Defaults to 3.
+
+        beta : :class:`float`, optional
+            Value of beta used in the in MC moves. Beta takes the
+            place of the inverse boltzmann temperature.
+            Defaults to 2.
+
+
+        """
+
+        self._output_dir = output_dir
+        self._step_size = step_size
+        self._target_bond_length = target_bond_length
+        self._num_steps = num_steps
+        self._bond_epsilon = bond_epsilon
+        self._nonbond_epsilon = nonbond_epsilon
+        self._nonbond_sigma = nonbond_sigma
+        self._nonbond_mu = nonbond_mu
+        self._beta = beta
+
+    def _get_bb_atom_ids(self, mol):
+
+        bb_ids = list(set([
+            i.get_building_block_id() for i in mol.get_atom_infos()
+        ]))
+        bb_atom_ids = {i: [] for i in bb_ids}
+        for i in mol.get_atom_infos():
+            bb_atom_ids[i.get_building_block_id()].append(
+                i.get_atom().get_id()
+            )
+
+        return bb_atom_ids
```

Comment:
I think we should file an issue to replace these functions with generators at some point. It would reduce memory and improve readability I think.

Reply:
This would require some significant rewriting and I am sure how much it would improve performance. Happy to write an issue.

Reply:
Yeah just stick it as a general improvement issue for now. I think there are a few functions that can be made into generators.